### PR TITLE
fix: successful runs fail after lease cleanup

### PR DESCRIPTION
**Pull request by excelsier**
## What Problem This Solves

Fixes an issue where a successful `crabbox run` could return exit code 7 when lease teardown killed an in-flight workspace-owner renewal after the command and lease cleanup had already succeeded.

## Why This Change Was Made

Provider teardown and the renewal loop previously raced. Cleanup now uses an explicit provider-neutral release lifecycle:

- Apple-container and Azure explicitly opt into quiescing normal renewal, preserving any failure observed against the live lease, and synchronously renewing ownership for a 10-minute release grace period before deletion;
- coordinator-backed leases always use the fence because their release endpoint is destructive (`delete=true`) for every coordinator-supported provider;
- retained stop/pause releases use a bounded short fence, preserving serialization through provider shutdown without leaving the destructive ten-minute fence; reachable `keep` releases retain the historical post-release remote owner cleanup;
- reachable/static releases keep the historical post-release remote owner cleanup;
- external adapters default to destructive grace fencing; weaker reachable or retained modes require an explicit trusted capability that is bound into routing/scope and cannot be enabled by repository config;
- provider cleanup plans bind a whole-release timeout and a fence of at least that timeout plus 30 seconds; Hostinger covers its two-minute stop wait, GitHub Codespaces covers its complete API sequence, and NVIDIA Brev covers its twelve-minute delete poll plus pre-delete work;
- short-fence plans are rejected before teardown if their normalized fence can outlive the explicit-reuse acquisition window; the current five-minute short fence fits inside the five-minute-thirty-second wait budget, so immediate reuse waits through expiry instead of timing out;
- reachable/static providers preserve the historical post-release remote owner `RELEASE`; retained and destructive providers preserve their installed fence on either success or an ambiguous release failure;
- only after the required preparation succeeds may provider teardown begin, so teardown cannot kill an in-flight renewal and another client cannot recover a destructive workspace during retries.

If pre-release renewal fails, cleanup is skipped and the run fails closed. Unknown providers default to the destructive fence. GitHub Codespaces preflights the effective delete disposition once: dirty/unpushed work is retained behind the short fence, while a clean or uncertain delete becomes delete-or-fail so a later state change cannot silently turn a ten-minute delete fence into a successful retained stop.

## User Impact

Apple-container and Azure one-shot runs no longer report a teardown-manufactured false failure after completing successfully. Genuine pre-release renewal failures remain fatal, destructive cleanup retains exclusive workspace ownership, and other providers retain their prior release behavior.

## Evidence

- Reproduced twice on Azure one-shot runs: command exit `0`, timing status `succeeded`, `leaseStopped=true`, provider inventory empty, followed by `remote workspace owner renewal failed closed: signal: killed` and outer exit `7`.
- Added a blocked-renewal regression proving pre-release failures remain fatal.
- Added a takeover regression proving a competing owner cannot recover after the original short TTL while the release-grace fence is active.
- Added lifecycle regressions proving Namespace retained-stop uses the bounded short fence, Apple-container/Azure use destructive grace, takeover remains blocked, and pre-release renewal failures remain fatal.
- Added a coordinator-adapter regression that exercises the real release endpoint, verifies `delete=true`, and proves the grace fence survives destructive release; provider-spec coverage proves static SSH and Namespace are coordinator-disabled.
- Added concrete production-provider disposition coverage for destructive, retained-stop, and reachable releases, including Coder config/lease precedence.
- Kept provider-specific disposition in provider adapters and added separate, non-launderable provenance for the external cleanup capability.
- Added retained-stop concurrency regressions proving reacquire is blocked before and during provider shutdown, succeeds after the bounded short fence, and remains fenced on shutdown failure.
- Added Codespaces regressions for dirty-to-clean drift, clean-to-dirty drift, uncertain preflight, early release errors, config drift, and plan-cache cleanup.
- Added Hostinger and NVIDIA Brev duration contracts, plus a destructive-release-failure regression through the production cleanup helper.
- Added whole-operation timeout cancellation, failed owner-preparation plan finalization, oversized short-plan rejection, immediate reuse through the maximum short fence, and Vast `keep` owner-release regressions.
- Added Coder config-vs-persisted-label and Firecracker persisted-state-record regressions, so cleanup planning now uses the exact authority each provider release path executes; uncertain Firecracker state fails closed to grace.
- Independent adversarial review of exact head `7e3ad05d82c33bd9ae164c4b36ad8d9a66c81310` found no actionable P0/P1/P2 issues; the prior adversarial pass explicitly verified the original six lifecycle finding classes, and the final focused pass verified Coder/Firecracker release-authority parity.
- Focused lifecycle/provider tests and focused race tests passed on exact head `7e3ad05d82c33bd9ae164c4b36ad8d9a66c81310` across core, GitHub Codespaces, Hostinger, NVIDIA Brev, Vast, Coder, and Firecracker.
- Full `go test ./...` passed on the same exact head.
- Exact-head real workload proof at `7e3ad05d82c33bd9ae164c4b36ad8d9a66c81310`: the built binary ran Bookeper's full cold Apple-container gate for 14m54.446s, returned command and outer launcher exit `0`, reported `runStatus=succeeded`, `totalMs=907451`, `endToEndMs=919790`, and `leaseStopped=true` for lease `cbx_5e9bf67f8b0c`.
- A second exact-head real workload proof used Bookeper's optimized but complete canonical gate at repository head `ed1d2ceff2616a7d773911aaf493ecae14975ec2`: command `510.534s`, Crabbox total `523.151s`, end-to-end `535.503s`, outer launcher `543s`, exit `0`, `runStatus=succeeded`, and `leaseStopped=true` for lease `cbx_a9f9def2ca62`. The run exceeded the normal renewal interval, retrieved and hash-verified the complete 8.56 MB replay-failed backend phase log before teardown, and left no lease/container/VM residue.
- The exact-head workload completed after the normal renewal interval and exercised keep=false destructive teardown. Post-run patched and installed provider inventories were empty; coordinator active leases were zero; direct Apple inventory contained only the persistent BuildKit helper; Azure provider GC and leak alarm reported zero VMs/orphans.

### CODE REVIEW

**Comment by excelsier** [comment]
Addressed the ordering finding at e23bde562b02a0678840060c2df88f1d5df84ef1: renewal now quiesces before provider release begins, so a pre-release failure remains fatal and teardown cannot manufacture a late renewal failure. Focused race coverage is green. I am adding the requested live production-boundary proof before asking for re-review.

**Comment by excelsier** [comment]
Live after-fix proof on exact head `e23bde562b02a0678840060c2df88f1d5df84ef1`:

- real Apple-container one-shot ran `sleep 15` (past renewal interval), then printed `patched-boundary-ok`
- command exit `0`; outer CLI exit `0`
- `runStatus=succeeded`, `commandMs=15110`, `totalMs=17808`, `endToEndMs=30270`
- `leaseStopped=true`
- post-run provider list empty; only Apple container persistent BuildKit helper remained
- no `remote workspace owner renewal failed closed` / outer exit 7

The revised implementation no longer suppresses errors. It prevents the teardown race by quiescing renewal before provider release and preserves any error observed on the live lease. @clawsweeper re-review

**Comment by excelsier** [comment]
Addressed the retry-window P1 at `cfcee8d036c70e528e94b1d0cd7ce60a00abc2ae`. Cleanup now quiesces ordinary renewal, synchronously extends the workspace-owner fence to 10 minutes (above the coordinator ~320s retry budget), and only then begins provider deletion. The 30s preparation bound remains below the normal 45s TTL; a failed extension skips cleanup and fails closed. Added takeover-after-original-TTL and pre-release-failure regressions. Focused, focused-race, and `go test ./...` proof are green. I am running the final exact-head live Apple-container proof now. @clawsweeper re-review

**Comment by excelsier** [comment]
Final exact-head live proof on `cfcee8d036c70e528e94b1d0cd7ce60a00abc2ae`: built the CLI from that head and ran a real Apple-container one-shot for 15s (past the normal 10s renewal interval). It printed `final-release-fence-ok`; command and outer CLI exit were `0`; timing reported `runStatus=succeeded`, `commandMs=15102`, `totalMs=17066`, `endToEndMs=29494`, and `leaseStopped=true` for lease `cbx_164f17b06006`. Post-run patched provider list `[]`; installed coordinator list `[]`; direct container inventory contains only Apple’s persistent BuildKit helper; Bookeper Azure/provider leak alarm is clean at 0 leases / 0 Azure VMs. No late renewal rc7 and no competing-cleanup gap.

**Comment by excelsier** [comment]
Resolved the persistent-target P1 on exact head `05e658f688723000ddec92ea0c7491f979a0f12c`. A provider-neutral reachability policy now keeps normal renewal through provider release for persistent targets, then performs remote owner `RELEASE`; destructive providers retain the 10-minute fail-closed grace fence. Actual cleanup-composition tests prove immediate persistent-target reacquire succeeds and destructive-target takeover remains blocked. After rebasing onto current upstream `main`, focused race tests and `go test ./...` passed. Exact-head Apple one-shot also exited 0 with `runStatus=succeeded`, `leaseStopped=true`, and zero residue. @clawsweeper re-review

**Comment by excelsier** [comment]
@codex review

File: internal/cli/run.go [hunk 1]
```diff
@@ -3209,6 +3223,31 @@ func releaseLeaseConnectionCleanupSafe(backend SSHLeaseBackend) bool {
...
+}
+
+func prepareWorkspaceOwnerForLeaseRelease(ctx context.Context, owner *workspaceOwner, backend SSHLeaseBackend) (bool, error) {
+	preservesTarget := releaseLeasePreservesReachableTarget(backend)
+	if owner == nil || preservesTarget {
+		return preservesTarget, nil
+	}
+	// Finish the synchronous grace renewal before the normal 45-second owner can
+	// expire; only then may destructive provider teardown begin.
+	return false, owner.PrepareLeaseRelease(ctx, workspaceOwnerReleaseGraceTTL)
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/cli/run.go]: Fixed on exact head `8132b68ac754305197a5ca27786df5182880a489` by removing the default/reachability classification entirely. The 10-minute owner grace is now an explicit destructive-release capability implemented only by Apple-container and Azure, the two reproduced rc7 providers. Namespace and every unclassified/retained provider keep the legacy no-preparation + full remote `Close` path. Focused provider/race tests, independent exact-head review, `go test ./...`, and a real Apple one-shot are green; cleanup reported `leaseStopped=true` with zero residue.

**Comment by excelsier** [comment]
Exact narrowed fix is ready at `8132b68ac754305197a5ca27786df5182880a489`. Only Apple-container and Azure opt into destructive release fencing; static, retained, and unclassified providers retain legacy owner-close behavior. Full suite, focused race tests, independent review, and exact-head live Apple proof are green. Live timing: command 15.107s, total 17.206s, end-to-end 29.916s, outer exit 0, `leaseStopped=true`; provider/coordinator/Azure inventories are clean. @clawsweeper re-review

**Comment by excelsier** [comment]
@codex review

File: internal/cli/workspace_owner.go [hunk 2]
```diff
@@ -16,9 +16,12 @@ import (
 const (
 	workspaceOwnerTTL           = 45 * time.Second
 	workspaceOwnerRenewInterval = 10 * time.Second
-	workspaceOwnerWaitTimeout   = 2 * time.Minute
-	workspaceOwnerPollInterval  = time.Second
-	workspaceOwnerProgressEvery = 10 * time.Second
+	// releaseCoordinatorLease can spend about 320 seconds across five timed
+	// attempts and backoff. Keep the fence valid beyond that whole retry window.
+	workspaceOwnerReleaseGraceTTL = 10 * time.Minute
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/cli/workspace_owner.go:27]: Fixed on `8132b68ac754305197a5ca27786df5182880a489`. Static SSH no longer implements or needs a special reachability policy: the 10-minute grace fence is explicit opt-in and only Apple-container/Azure opt in. Static SSH follows the legacy path and performs full remote owner `Close`, so immediate reacquire remains covered by the focused lifecycle regression.

File: internal/cli/run.go [hunk 3]
```diff
@@ -954,6 +954,10 @@ func (a App) runCommandWithBenchmarkRecord(ctx context.Context, args []string, b
 				fmt.Fprintf(a.Stderr, "lease cleanup skipped while workspace child remains possible: %v\n", ownerErr)
 				return
 			}
+			// No renewal may remain in flight when provider teardown begins. This
+			// preserves genuine failures on the live lease and prevents teardown
+			// from manufacturing a late signal-killed renewal failure.
+			lifecycleOwner.StopRenewalBeforeLeaseRelease()
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/cli/run.go]: Fixed on `8132b68ac754305197a5ca27786df5182880a489`. Apple-container and Azure now renew a 10-minute owner fence before destructive release, so their cleanup cannot outlive the normal 45-second TTL. CoordinatorLeaseBackend is exclusive and does not acquire this workspace owner; all other unclassified/retained providers preserve legacy behavior rather than receiving a new long fence. The destructive slow-release takeover regression and exact live Apple proof are green.

**Comment by excelsier** [comment]
Maintainer note: the fork-triggered CI, Connector E2E, and release-check workflows are currently `action_required` and need repository approval. Exact head `8132b68ac754305197a5ca27786df5182880a489` has local `go test ./...`, focused race tests, independent review, and a real Apple-container cleanup proof green; all review threads are resolved. Please approve the queued workflow runs when convenient.

File: internal/cli/run.go [hunk 4]
```diff
@@ -3209,6 +3223,36 @@ func releaseLeaseConnectionCleanupSafe(backend SSHLeaseBackend) bool {
 	return !ok || policy.ReleaseLeaseConnectionCleanupSafe()
 }
 
+func releaseLeaseNeedsOwnerGraceFence(backend SSHLeaseBackend, lease LeaseTarget) bool {
+	policy, ok := backend.(ReleaseLeaseOwnerGraceFencePolicy)
+	return ok && policy.ReleaseLeaseNeedsOwnerGraceFence(lease)
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/cli/run.go]: Fixed in `2d1d94cb`. The coordinator adapter now always opts into the owner grace fence because its release endpoint is destructive (`delete=true`) for every coordinator-supported provider. `TestWorkspaceOwnerCoordinatorReleaseDeletePathUsesGraceFence` exercises the actual endpoint and asserts the destructive request plus retained fence; the provider-spec boundary test proves static SSH and Namespace are `CoordinatorNever`. Focused, race, and full `internal/cli` tests pass; independent exact-head re-review found no P0/P1/P2.

**Comment by excelsier** [comment]
@clawsweeper re-review

**Comment by excelsier** [comment]
@codex review

File: internal/cli/provider_coordinator.go [hunk 5]
```diff
@@ -49,6 +49,11 @@ func (b *coordinatorLeaseBackend) RebindResolvedLeaseTarget(target *LeaseTarget,
 	return nil
 }
 
+func (b *coordinatorLeaseBackend) ReleaseLeaseNeedsOwnerGraceFence(lease LeaseTarget) bool {
+	policy, ok := b.direct.(ReleaseLeaseOwnerGraceFencePolicy)
+	return ok && policy.ReleaseLeaseNeedsOwnerGraceFence(lease)
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/cli/provider_coordinator.go]: Fixed in `2d1d94cb`. `coordinatorLeaseBackend.ReleaseLeaseNeedsOwnerGraceFence` now returns true for every coordinator-backed lease because the adapter always releases with `delete=true`. The regression calls the real coordinator endpoint and verifies the destructive request plus retained fence; the spec-boundary test covers AWS/GCP/Hetzner/Daytona/Azure as coordinator-supported and static SSH/Namespace as coordinator-disabled. Full CLI, focused race, and independent exact-head review are green.

**Comment by excelsier** [comment]
Maintainer update: current exact head is `2d1d94cbeb1f7a5debf29371b92b78007d84aeff`. The late coordinator P1 is fixed: all coordinator releases now fence because the endpoint is destructive (`delete=true`), with actual endpoint, provider-boundary, race, full CLI, and independent review proof green; unresolved review threads are zero. The current-head CI, Connector E2E, and release-check runs remain `action_required` pending repository approval.

**Comment by excelsier** [comment]
@clawsweeper re-review

**Comment by excelsier** [comment]
@clawsweeper re-review

**Comment by excelsier** [comment]
@codex review

**Comment by excelsier** [comment]
@clawsweeper re-review

**Comment by excelsier** [comment]
@codex review

File: internal/providers/githubcodespaces/backend.go [hunk 6]
```diff
@@ -766,6 +768,10 @@ func (b *backend) RetainLeaseClaimAfterRelease(lease LeaseTarget) bool {
 	return !githubCodespacesDeleteOnRelease(lease, b.cfg)
 }
 
+func (b *backend) ReleaseLeaseOwnerCleanupMode(lease LeaseTarget) core.ReleaseLeaseOwnerCleanupMode {
+	return core.ReleaseLeaseOwnerCleanupModeForConfig(providerName, b.cfg, lease)
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/providers/githubcodespaces/backend.go]: Closed at exact head 7bead0f4. Codespaces now preflights the effective resource state, caches forceStop versus forceDelete, consumes that decision once, and treats a later unsafe delete as fail-closed unless retained stop was explicitly planned. Dirty/unpushed work therefore receives the short retained fence; clean or uncertain delete is delete-or-fail. Drift, uncertainty, and cache-finalization regressions pass.

File: internal/cli/run.go [hunk 7]
```diff
@@ -566,19 +566,26 @@ func (a App) runCommandWithBenchmarkRecord(ctx context.Context, args []string, b
...
+			if closeErr != nil {
 				runFailure = recordRunFailure(&runFailure, closeErr)
 				err = errors.Join(err, closeErr)
-				fmt.Fprintf(a.Stderr, "warning: workspace owner stopped with released lease after renewal failure: %v\n", closeErr)
+				fmt.Fprintf(a.Stderr, "warning: workspace owner close after lease release failed: %v\n", closeErr)
 			}
 			return
 		}
+		if lifecycleOwnerCleanupMode == ReleaseLeaseOwnerCleanupShortFence {
+			return
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/cli/run.go]: Closed at exact head 7bead0f4. Both short-fence and grace-fence modes retain the renewed marker when provider release fails; only reachable after-provider-release cleanup attempts remote owner close. Whole release is bounded and every fenced plan normalizes to ReleaseTimeout plus a 30-second safety margin. The destructive-release-failure regression exercises the production cleanup helper.

File: internal/providers/hostinger/backend.go [hunk 8]
```diff
@@ -575,6 +577,10 @@ func (b *leaseBackend) RetainLeaseClaimAfterRelease(LeaseTarget) bool {
 	return true
 }
 
+func (b *leaseBackend) ReleaseLeaseOwnerCleanupMode(LeaseTarget) core.ReleaseLeaseOwnerCleanupMode {
+	return core.ReleaseLeaseOwnerCleanupShortFence
+}
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/providers/hostinger/backend.go:583]: Closed at exact head 7bead0f4. Hostinger declares a 2m30s whole-release timeout for its 2m stop wait. Core applies that bound around refresh, connection cleanup, and provider ReleaseLease, while the normalized short fence is 5m and remains below the 5m30s explicit-reuse wait. Provider and core timeout regressions pass.

File: internal/providers/nvidiabrev/backend.go [hunk 9]
```diff
@@ -288,6 +290,13 @@ func (b *nvidiaBrevBackend) RetainLeaseClaimAfterRelease(lease LeaseTarget) bool
 	return true
 }
 
+func (b *nvidiaBrevBackend) ReleaseLeaseOwnerCleanupMode(lease LeaseTarget) core.ReleaseLeaseOwnerCleanupMode {
+	if b.releaseAction(lease.Server.Labels) == "stop" {
+		return core.ReleaseLeaseOwnerCleanupShortFence
+	}
+	return core.ReleaseLeaseOwnerCleanupGraceFence
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/providers/nvidiabrev/backend.go:298]: Closed at exact head 7bead0f4. NVIDIA Brev declares a 15m whole-release timeout covering its 12m delete poll plus pre-delete work; core bounds the entire release call, and the grace fence normalizes to 15m30s. Focused, race, and full Go suites pass.

File: internal/providers/vast/backend.go [hunk 10]
```diff
@@ -882,6 +882,15 @@ func effectiveVastReleaseAction(cfg core.Config, labels map[string]string) strin
 	return normalizeVastReleaseAction(firstNonBlank(labels[vastReleaseActionLabel], cfg.Vast.ReleaseAction))
 }
 
+func (b *backend) ReleaseLeaseOwnerCleanupMode(lease core.LeaseTarget) core.ReleaseLeaseOwnerCleanupMode {
+	switch effectiveVastReleaseAction(b.cfg, lease.Server.Labels) {
+	case "keep", "stop":
+		return core.ReleaseLeaseOwnerCleanupShortFence
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/providers/vast/backend.go]: Closed at exact head 7bead0f4. Vast keep now maps to after-provider-release; its no-op provider release is followed by the normal remote owner RELEASE, so it is immediately reusable. Vast stop remains short-fenced and destroy remains grace-fenced. Persisted and explicit configuration precedence regressions pass.

File: internal/cli/workspace_owner.go [hunk 11]
```diff
@@ -16,9 +16,16 @@ import (
...
-	workspaceOwnerPollInterval  = time.Second
-	workspaceOwnerProgressEvery = 10 * time.Second
+	// Retained stop/pause/keep releases can make SSH unreachable before a remote
+	// RELEASE is possible. Keep the marker through bounded provider cleanup, then
+	// let it expire promptly enough for reuse.
+	workspaceOwnerReleaseShortTTL = 3 * time.Minute
+	// releaseCoordinatorLease can spend about 320 seconds across five timed
+	// attempts and backoff. Keep the fence valid beyond that whole retry window.
+	workspaceOwnerReleaseGraceTTL = 10 * time.Minute
+	workspaceOwnerWaitTimeout     = 2 * time.Minute
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/cli/workspace_owner.go]: Closed at exact head 7bead0f4. The current short fence is 5m and explicit reuse waits 5m30s. Production validation rejects any normalized short plan whose fence is at or beyond that wait budget and finalizes cached plan state before skipping provider teardown. An immediate-reacquire regression waits through the maximum short fence and succeeds; oversized plans fail closed.

**Comment by excelsier** [comment]
Exact head `7bead0f4d87380052b26920a3a3c2b5aa7232c2e` closes all six lifecycle threads. Focused, race, full Go, independent adversarial review, and live Apple one-shot evidence are in the PR body; post-run inventories are clean. @clawsweeper re-review

**Comment by excelsier** [comment]
@codex review

File: internal/providers/coder/backend.go [hunk 12]
```diff
@@ -364,6 +367,10 @@ func (b *coderLeaseBackend) ReleaseLeaseMessage(lease LeaseTarget) string {
 	return fmt.Sprintf("%s coder workspace lease=%s workspace=%s", action, lease.LeaseID, lease.Server.Name)
 }
 
+func (b *coderLeaseBackend) ReleaseLeaseOwnerCleanupMode(lease LeaseTarget) core.ReleaseLeaseOwnerCleanupMode {
+	return shared.DeleteOnReleaseOwnerCleanupMode(shared.DeleteOnReleaseFromLease(b.cfg.Coder.DeleteOnRelease, core.DeleteOnReleaseExplicit(b.cfg, coderProvider), lease))
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/providers/coder/backend.go]: Closed at exact head 7e3ad05d. Coder cleanup mode now uses `b.cfg.Coder.DeleteOnRelease`, the same current configuration source that `ReleaseLease`/`releaseWorkspace` executes. Persisted labels no longer override the actual command. Regression covers both drift directions: current stop with persisted delete is short-fenced; current delete with persisted stop is grace-fenced.

File: internal/providers/firecracker/backend.go [hunk 13]
```diff
@@ -409,6 +409,11 @@ func (b *backend) ReleaseLease(ctx context.Context, req ReleaseLeaseRequest) err
 	return nil
 }
 
+func (b *backend) ReleaseLeaseOwnerCleanupMode(lease LeaseTarget) core.ReleaseLeaseOwnerCleanupMode {
+	cfg := b.configForRun()
+	return shared.DeleteOnReleaseOwnerCleanupMode(shared.DeleteOnReleaseFromLease(cfg.Firecracker.DeleteOnRelease, core.DeleteOnReleaseExplicit(cfg, providerName), lease))
```
[the remark this answers is not in the document]
Comment (excelsier) [reply internal/providers/firecracker/backend.go:414]: Closed at exact head 7e3ad05d. Firecracker now implements a cleanup planner that loads the same persisted state record used by `ReleaseLease` and derives the mode from `record.DeleteOnRelease`. Missing/uncertain state fails closed to a grace fence. Regressions cover current-delete/persisted-stop, current-stop/persisted-delete, and missing state.

**Comment by excelsier** [comment]
Exact head `7e3ad05d82c33bd9ae164c4b36ad8d9a66c81310` aligns Coder and Firecracker cleanup planning with the exact release authority each provider executes. Focused, race, full Go, and independent exact-head review are green; unresolved threads are zero. @clawsweeper re-review

**Comment by excelsier** [comment]
@codex review

**Comment by excelsier** [comment]
Maintainer-ready exact-head update: Codex reviewed `7e3ad05d82` clean, unresolved threads are zero, and the PR body now includes a fully hydrated real workload run using the exact built head. Bookeper full cold Apple gate ran 14m54.446, command and outer launcher exited 0, `runStatus=succeeded`, `leaseStopped=true`, and all provider/coordinator/Azure/GC inventories were clean. This also supplies the hydrated exact-head lifecycle proof requested by ClawSweeper. Fork CI/Connector/Release/Docs workflows remain `action_required` pending repository approval.

**Comment by excelsier** [comment]
Additional exact-head lifecycle proof is now in the PR body: the optimized-but-complete Bookeper canonical gate ran 9m03 outer / 8m30 command on the exact `7e3ad05d82` Crabbox binary, exited 0, retrieved and hash-verified the replay-failed 8.56 MB backend phase log before teardown, reported `leaseStopped=true`, and left zero leases/containers/Azure VMs/GC candidates. This is a second real workload beyond the earlier 14m54 run and exercises the normal renewal interval plus destructive keep=false cleanup.

**Comment by steipete** [comment]
Thanks @excelsier — the lease-cleanup failure you targeted is real, and parts of this were locally verified sound. Closing though: review found four independent blockers (three were repairable, the fourth surfaced on the full-branch gate), and at that point the change had stopped converging — continuing meant an unbounded rewrite of the architecture in this PR's shape.

The recommended path, which we'd welcome: rewrite from current main around an explicit owner-preparation state (`unprepared` / `fenced` / `reachable-release`) so deferred cleanup keys off what actually succeeded rather than the requested plan mode; keep provider classification behind adapters; and add one end-to-end composition test per failure edge before any broad provider sweep.
